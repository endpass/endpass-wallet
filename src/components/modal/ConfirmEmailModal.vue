<template>
  <v-modal
    class="is-dark"
    @close="handleClose"
  >
    <template slot="header">Check Your Email</template>

    <p class="subtitle">We just sent you an email with a confirmation link.
    Click it to instantly log in to your wallet!</p>
  </v-modal>
</template>

<script>
export default {
  name: 'ConfirmEmailModal',
  methods: {
    handleClose() {
      this.$emit('close');
    },
  },
};
</script>
