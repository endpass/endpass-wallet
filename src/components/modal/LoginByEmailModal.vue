<template>
  <v-modal
    class="is-dark"
    @close="handleClose"
  >
    <template slot="header">Get Started</template>

    <p class="subtitle">Please enter your email address below to access your
    wallet or create a new one.</p>

    <v-form
      id="loginByEmail"
      v-model="isFormValid"
      @submit="handleSubmit"
    >
      <v-input
        v-model="email"
        :disabled="isLoading"
        label="Email"
        help="Your email address may be used to help recover your
               wallet in case you lose access."
        name="email"
        validator="required|email"
        placeholder="Your email"
      />

      <v-select
        :disabled="isLoading"
        :options="availableIdentityServerTypes"
        v-model="currentIdentityServerType"
        label="Identity Server"
        name="currentIdentityServerType"
      />

      <v-input
        v-if="selectedCustomIdentityServerType"
        id="customIdentityServer"
        v-model="customIdentityServer"
        :disabled="isLoading"
        label="Custom Identity Server"
        name="customIdentityServer"
        validator="required|url:require_protocol:true"
        placeholder="Custom Identity Server"
      />

      <v-checkbox v-model="termsAccepted">
        I accept the <a
          href="https://endpass.com/terms/"
          target="_blank"
        >
          Terms of Service
        </a>
        and
        <a
          href="https://endpass.com/privacy/"
          target="_blank"
        >
          Privacy Policy
        </a>.
      </v-checkbox>
    </v-form>
    <div
      slot="footer"
      class="buttons"
    >
      <v-button
        :disabled="!termsAccepted || !isFormValid"
        :loading="isLoading"
        class-name="is-primary is-medium"
        form="loginByEmail"
        data-test="submit-login"
      >
        Continue
      </v-button>
    </div>
  </v-modal>
</template>

<script>
import VModal from '@/components/ui/VModal';
import VForm from '@/components/ui/form/VForm';
import VInput from '@/components/ui/form/VInput';
import VButton from '@/components/ui/form/VButton';
import VCheckbox from '@/components/ui/form/VCheckbox';
import VSelect from '@/components/ui/form/VSelect';
import { IDENTITY_MODE } from '@/constants';

const availableIdentityServerTypes = [
  { text: 'Endpass', val: IDENTITY_MODE.DEFAULT },
  { text: 'Local Storage', val: IDENTITY_MODE.LOCAL },
  { text: 'Custom server', val: IDENTITY_MODE.CUSTOM },
];

export default {
  name: 'LoginByEmailModal',
  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    email: '',
    termsAccepted: true,
    availableIdentityServerTypes,
    currentIdentityServerType: availableIdentityServerTypes[0].val,
    customIdentityServer: null,
    isFormValid: false,
  }),
  computed: {
    selectedCustomIdentityServerType() {
      return this.currentIdentityServerType === IDENTITY_MODE.CUSTOM;
    },
  },
  methods: {
    handleSubmit() {
      this.$ga.event({
        eventCategory: 'onboarding',
        eventAction: 'submit_email',
      });

      const { email } = this;
      const mode = {
        type: this.currentIdentityServerType,
        serverUrl: this.selectedCustomIdentityServerType
          ? this.customIdentityServer
          : undefined,
      };

      this.$emit('confirm', { email, mode });
    },
    handleClose() {
      this.$emit('close');
    },
  },
  components: {
    VModal,
    VForm,
    VInput,
    VButton,
    VCheckbox,
    VSelect,
  },
};
</script>
