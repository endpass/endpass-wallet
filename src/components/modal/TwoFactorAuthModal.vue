<template lang="html">
  <v-modal
    class="is-dark"
    @close="close"
  >
    <template slot="header">Two Factor Authentication</template>
    <div v-if="secret">
      <p class="lead">To protect your account, we strongly recommend enabling
      two factor authentication. Once enabled, you will be able to log in by
      entering a code generated by an authenticator app.</p>

      <p class="lead">Scan the QR code below into <strong>Google
      Authenticator</strong>. If you are unable to scan the code, you can
      also enter the secret manually.</p>

    </div>
    <v-form
      id="twoFactorAuth"
      @submit="confirm"
    >
      <div
        v-if="secret && email"
        class="field"
      >
        <div class="has-text-centered">
          <img :src="qrCodeSrc">
        </div>
        <p class="subtitle">Write down the recovery secret below:</p>
        <label class="label">Secret</label>
        <p class="subtitle twofa-secret">
          <strong>{{ secret }}</strong>
        </p>
      </div>
      <p class="subtitle">Enter the 6 digit code generated by your
      authenticator app.</p>
      <v-input
        v-model="code"
        type="number"
        size="6"
        label="Verification Code"
        name="verificationCode"
        validator="required|digits:6"
      />
    </v-form>
    <div
      slot="footer"
      class="buttons"
    >
      <v-button
        :loading="isLoading"
        form="twoFactorAuth"
        class-name="is-primary is-medium"
      >
        Verify
      </v-button>
    </div>
  </v-modal>
</template>

<script>
import VModal from '@/components/ui/VModal';
import VForm from '@/components/ui/form/VForm';
import VInput from '@/components/ui/form/VInput';
import VButton from '@/components/ui/form/VButton';

export default {
  // TODO: must be in PascalCase, breaks tests
  name: 'two-factor-auth-modal',
  props: {
    secret: String,
    email: String,
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      code: null,
    };
  },
  computed: {
    qrCodeSrc() {
      const otpAuthUri = `otpauth://totp/Endpass:${this.email}?secret=${
        this.secret
      }`;

      return `https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=${otpAuthUri}`;
    },
  },
  methods: {
    confirm() {
      this.$emit('confirm', this.code);
    },
    close() {
      this.$emit('close');
    },
  },
  components: {
    VModal,
    VForm,
    VInput,
    VButton,
  },
};
</script>
