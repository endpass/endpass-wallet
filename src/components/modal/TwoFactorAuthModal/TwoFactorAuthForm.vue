<template>
  <v-form
    id="twoFactorAuthForm"
    :is-form-valid="isFormValid"
    @submit.native.prevent="onFormSubmit"
  >
    <div
      v-if="isShowQRCode"
      class="two-factor-auth-form-section"
      data-test="two-factor-modal-qr-code"
    >
      <div class="has-text-centered">
        <img :src="qrCodeSrc">
      </div>
      <label class="subtitle">
        {{ $t('components.twoFactorAuthModal.writeDownCode') }}
      </label>
      <label class="label">
        {{ $t('components.twoFactorAuthModal.secret') }}
      </label>
      <p class="subtitle twofa-secret">
        <strong>{{ secret }}</strong>
      </p>
    </div>
    <div class="two-factor-auth-form-section">
      <label class="subtitle">
        {{ $t('components.twoFactorAuthModal.enterCode') }}
      </label>
      <v-input
        v-model="code"
        v-validate="'required|digits:6'"
        :error="errors.first('code')"
        type="number"
        size="6"
        :label="$t('components.twoFactorAuthModal.otpCode')"
        name="code"
        data-vv-name="code"
        data-test="input-two-auth-code"
        autofocus
        @input="handleInput"
      />
    </div>
    <v-button
      :loading="isLoading"
      :disabled="!isFormValid"
      type="submit"
      class-name="is-primary is-medium"
      data-test="submit-two-auth-modal"
    >
      {{ $t('global.verify') }}
    </v-button>
  </v-form>
</template>

<script>
import formMixin from '@/mixins/form';

export default {
  name: 'TwoFactorAuthForm',

  props: {
    secret: {
      type: String,
      default: '',
    },
    email: {
      type: String,
      default: '',
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    code: '',
  }),

  computed: {
    isShowQRCode() {
      return this.email && this.secret;
    },

    qrCodeSrc() {
      const otpAuthUri = `otpauth://totp/Endpass:${this.email}?issuer=Endpass&secret=${this.secret}`;

      return `https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=${encodeURIComponent(
        otpAuthUri,
      )}`;
    },
  },

  methods: {
    onFormSubmit() {
      if (!this.isFormValid) return;

      this.$emit('submit', this.code);
    },
  },

  mixins: [formMixin],
};
</script>

<style lang="scss">
.two-factor-auth-form-section {
  margin-bottom: 30px;
}
</style>
