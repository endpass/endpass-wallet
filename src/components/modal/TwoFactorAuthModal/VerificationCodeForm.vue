<template>
  <v-form
    id="verificationCode"
    :is-form-valid="isFormValid"
    @submit="onFormSubmit"
  >
    <div class="verification-code-form-section">
      <label class="subtitle">
        {{ $t('components.twoFactorAuthModal.enterVerificationCode') }}
      </label>
      <v-input
        v-model="verificationCode"
        v-validate="'required|digits:6'"
        :error="errors.first('verificationCode')"
        type="number"
        size="6"
        :label="$t('components.twoFactorAuthModal.verificationCode')"
        name="verificationCode"
        data-vv-name="verificationCode"
        data-test="input-two-auth-verification-code"
        autofocus
      />
    </div>
    <div class="verification-code-form-section v-text-center">
      <span>{{ $t('components.twoFactorAuthModal.didNotGetCode') }}</span>
      <a class="link" @click="onCodeRequest">
        {{ $t('components.twoFactorAuthModal.resendCode') }}
      </a>
    </div>
    <v-button
      :loading="isLoading"
      :disabled="!isFormValid"
      class-name="is-primary is-medium"
      data-test="submit-two-auth-modal"
    >
      {{ $t('global.continue') }}
    </v-button>
  </v-form>
</template>

<script>
import formMixin from '@/mixins/form';

export default {
  name: 'VerificationCodeForm',

  props: {
    email: {
      type: String,
      default: '',
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    verificationCode: '',
  }),

  methods: {
    onFormSubmit() {
      if (!this.isFormValid) return;

      this.$emit('submit', this.verificationCode);
    },

    onCodeRequest() {
      this.$emit('request-code');
    },
  },

  mixins: [formMixin],
};
</script>

<style lang="scss">
.verification-code-form-section {
  margin-bottom: 30px;
}
</style>
