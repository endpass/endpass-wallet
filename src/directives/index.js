import dynamicEvents from './dynamicEvents';

export default Vue => {
  dynamicEvents(Vue);
};
