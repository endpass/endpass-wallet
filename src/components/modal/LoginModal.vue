<template>
  <component
    v-dynamic-events="['confirm', 'close']"
    :is="currentModal"
    :is-loading="isLoading"
    class="login-modal"
    data-test="login-modal"
  />
</template>

<script>
import { mapActions } from 'vuex';

import LoginByEmailModal from '@/components/modal/LoginByEmailModal';
import ConfirmEmailModal from '@/components/modal/ConfirmEmailModal';
import TwoFactorAuthModal from '@/components/modal/TwoFactorAuthModal';
import error from '@/mixins/error';

export default {
  name: 'LoginModal',
  data: () => ({
    currentModal: LoginByEmailModal.name,
    isLoading: false,
    email: null,
  }),
  methods: {
    ...mapActions('user', ['login', 'loginViaOTP']),
    handleLoginByEmailModalConfirm(email) {
      this.isLoading = true;

      return this.login(email)
        .then(challengeType => {
          if (challengeType === 'otp') {
            this.email = email;
            this.currentModal = TwoFactorAuthModal.name;
            this.isLoading = false;
          } else if (challengeType === 'email_link') {
            this.currentModal = ConfirmEmailModal.name;
            this.isLoading = false;
          } else {
            this.handleSuccessfulLogin();
          }
        })
        .catch(this.handleFailedLogin);
    },
    handleTwoFactorAuthModalConfirm(code) {
      this.isLoading = true;
      const {
        email,
        handleFailedLogin,
        handleSuccessfulLogin,
        loginViaOTP,
      } = this;

      return loginViaOTP({ code, email })
        .then(handleSuccessfulLogin)
        .then(this.redirectPage)
        .catch(handleFailedLogin);
    },
    handleClose() {
      this.close();
    },
    handleSuccessfulLogin() {
      this.isLoading = false;
      this.close();

      this.$notify({
        title: 'Success',
        type: 'is-info',
        text: 'Logged In',
      });
    },
    handleFailedLogin(error) {
      this.isLoading = false;
      this.emitError(error);
    },
    close() {
      this.$emit('close');
    },
    redirectPage() {
      const regirectUri = this.$route.query.redirect_uri;

      if (regirectUri) {
        this.$router.push({
          path: regirectUri,
        });
      }
    },
  },
  mixins: [error],
  components: {
    LoginByEmailModal,
    ConfirmEmailModal,
    TwoFactorAuthModal,
  },
};
</script>
