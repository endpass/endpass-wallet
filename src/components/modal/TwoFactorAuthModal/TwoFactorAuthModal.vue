<template>
  <v-modal class="is-dark two-factor-auth-modal" @close="close">
    <template slot="header">
      {{ $t('components.twoFactorAuthModal.header') }}
    </template>
    <two-factor-auth-form
      :secret="secret"
      :email="email"
      :is-loading="isLoading"
      @submit="onFormSubmit"
    />
  </v-modal>
</template>

<script>
import TwoFactorAuthForm from './TwoFactorAuthForm';

export default {
  name: 'TwoFactorAuthModal',

  props: {
    secret: {
      type: String,
      default: '',
    },
    email: {
      type: String,
      default: '',
    },
    isLoading: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    onFormSubmit({ code, verificationCode }) {
      this.$emit('confirm', {
        verificationCode,
        code,
      });
    },

    close() {
      this.$emit('close');
    },
  },

  components: {
    TwoFactorAuthForm,
  },
};
</script>

<style lang="scss">
.two-factor-auth-modal-section {
  margin-bottom: 30px;
}
</style>
