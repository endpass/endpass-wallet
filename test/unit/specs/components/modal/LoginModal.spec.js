import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Notifications from 'vue-notification';
import directives from '@/directives';

import LoginModal from '@/components/modal/LoginModal';
import LoginByEmailModal from '@/components/modal/LoginByEmailModal';
import ConfirmEmailModal from '@/components/modal/ConfirmEmailModal';
import TwoFactorAuthModal from '@/components/modal/TwoFactorAuthModal';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Notifications);
localVue.use(directives);

jest.mock('@/components/modal/LoginByEmailModal', () => ({
  name: 'login-by-email-modal',
}));

jest.mock('@/components/modal/ConfirmEmailModal', () => ({
  name: 'confirm-email-modal',
}));

jest.mock('@/components/modal/TwoFactorAuthModal', () => ({
  name: 'two-factor-auth-modal',
}));

describe('LoginModal', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(LoginModal, {
      localVue,
      sync: false,
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('LoginModal');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render LoginByEmailModal component', async () => {
      expect.assertions(1);

      wrapper.setData({
        currentModal: LoginByEmailModal.name,
      });

      wrapper.vm.$forceUpdate();
      await wrapper.vm.$nextTick();

      expect(wrapper.is(`${LoginByEmailModal.name}-stub`)).toBeTruthy();
    });

    it('should render ConfirmEmailModal component', async () => {
      expect.assertions(1);

      wrapper.setData({
        currentModal: ConfirmEmailModal.name,
      });

      wrapper.vm.$forceUpdate();
      await wrapper.vm.$nextTick();

      expect(wrapper.is(`${ConfirmEmailModal.name}-stub`)).toBeTruthy();
    });

    it('should render TwoFactorAuthModal component', async () => {
      expect.assertions(1);

      wrapper.setData({
        currentModal: TwoFactorAuthModal.name,
      });

      wrapper.vm.$forceUpdate();
      await wrapper.vm.$nextTick();

      expect(wrapper.is(`${TwoFactorAuthModal.name}-stub`)).toBeTruthy();
    });
  });

  describe('data', () => {
    describe('isLoading', () => {
      it('should correctly change "loading" property', async () => {
        wrapper.setData({
          isLoading: false,
        });

        wrapper.vm.$forceUpdate();
        await wrapper.vm.$nextTick();

        expect(wrapper.attributes()['is-loading']).toBeUndefined();

        wrapper.setData({
          isLoading: true,
        });

        wrapper.vm.$forceUpdate();
        await wrapper.vm.$nextTick();

        expect(wrapper.attributes()['is-loading']).toBeTruthy();
      });
    });
  });

  describe('methods', () => {
    const actions = {
      login: jest.fn(),
      loginViaOTP: jest.fn(),
    };
    let wrapper;

    beforeEach(() => {
      const $router = {
        push: jest.fn(),
      };
      const $route = {
        query: {},
      };
      const $ga = { event: jest.fn() };

      const store = new Vuex.Store({
        modules: {
          user: {
            namespaced: true,
            actions,
          },
          errors: {
            namespaced: true,
            actions: {
              emitError: jest.fn(),
            },
          },
        },
        actions: {
          init: jest.fn().mockResolvedValue(),
        },
      });

      wrapper = shallowMount(LoginModal, {
        store,
        localVue,
        mocks: { $route, $router, $ga },
      });
    });

    afterEach(() => {
      actions.login.mockClear();
      actions.loginViaOTP.mockClear();
    });

    describe('handleLoginByEmailModalConfirm', () => {
      const email = 'email';

      beforeEach(() => {
        jest.spyOn(wrapper.vm, '$notify');
      });

      it('should login user via emailLink challenge type', async () => {
        expect.assertions(3);

        actions.login.mockResolvedValueOnce('emailLink');

        await wrapper.vm.handleLoginByEmailModalConfirm(email);

        expect(wrapper.vm.isLoading).toBeFalsy();
        expect(wrapper.vm.currentModal).toBe(ConfirmEmailModal.name);
        expect(wrapper.emitted().close).toBeUndefined();
      });

      it('should login user via OTP challenge type', async () => {
        expect.assertions(4);

        actions.login.mockResolvedValueOnce('otp');

        await wrapper.vm.handleLoginByEmailModalConfirm(email);

        expect(wrapper.vm.isLoading).toBeFalsy();
        expect(wrapper.vm.currentModal).toBe(TwoFactorAuthModal.name);
        expect(wrapper.vm.$notify).toHaveBeenCalledTimes(0);
        expect(wrapper.emitted().close).toBeUndefined();
      });

      it('should not login user', async () => {
        expect.assertions(5);

        const error = {};

        jest.spyOn(wrapper.vm, 'emitError');
        actions.login.mockRejectedValueOnce(error);

        await wrapper.vm.handleLoginByEmailModalConfirm(email);

        expect(wrapper.vm.isLoading).toBeFalsy();
        expect(wrapper.vm.currentModal).toBe(LoginByEmailModal.name);
        expect(wrapper.vm.emitError).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.emitError).toHaveBeenCalledWith(error);
        expect(wrapper.emitted().close).toBeUndefined();
      });
    });

    describe('handleTwoFactorAuthModalConfirm', () => {
      const code = '111';

      beforeEach(() => {
        jest.spyOn(wrapper.vm, '$notify');

        wrapper.setData({
          currentModal: TwoFactorAuthModal.name,
        });
      });

      it('should login user via OTP challenge type', async () => {
        await wrapper.vm.handleTwoFactorAuthModalConfirm(code);

        expect(wrapper.vm.isLoading).toBeFalsy();
        expect(wrapper.vm.$notify).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$notify).toHaveBeenCalledWith({
          title: 'Success',
          text: 'Logged In',
          type: 'is-info',
        });
        expect(wrapper.emitted().close).toEqual([[]]);
      });

      it('should not login user', async () => {
        const error = {};

        jest.spyOn(wrapper.vm, 'emitError');
        actions.loginViaOTP.mockRejectedValueOnce(error);

        await wrapper.vm.handleTwoFactorAuthModalConfirm(code);

        expect(wrapper.vm.isLoading).toBeFalsy();
        expect(wrapper.vm.emitError).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.emitError).toHaveBeenCalledWith(error);
        expect(wrapper.emitted().close).toBeUndefined();
      });

      it('should redirect after login', async () => {
        const regirectUri = '/some-page';

        wrapper.vm.$route.query.redirect_uri = regirectUri;

        await wrapper.vm.handleTwoFactorAuthModalConfirm(code);

        expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$router.push).toHaveBeenCalledWith({
          path: regirectUri,
        });
      });

      it('should not redirect after login', async () => {
        await wrapper.vm.handleTwoFactorAuthModalConfirm(code);

        expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(0);
      });

      it('should reload data after successfully login', async () => {
        expect.assertions(1);

        wrapper.setMethods({
          reloadData: jest.fn().mockResolvedValue(),
        });

        await wrapper.vm.handleTwoFactorAuthModalConfirm(code);

        expect(wrapper.vm.reloadData).toHaveBeenCalledTimes(1);
      });
    });
  });
});
