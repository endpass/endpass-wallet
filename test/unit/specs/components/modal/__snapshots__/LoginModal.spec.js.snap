// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginModal render should render component 1`] = `
<login-by-email-modal-stub
  class="login-modal"
  data-test="login-modal"
/>
`;
